Source/Dafny/Resolver.cs

```diff
@@ -4546,13 +4546,7 @@ public class ModuleBindings {
       Contract.Ensures(!Contract.ValueAtReturn(out isRoot) || Contract.ValueAtReturn(out headIsRoot)); // isRoot ==> headIsRoot
```

Comment:
Could you add a description of what this method does or in which context is is being used? I don't know what "CheckEnds" means, which "End" it is refered to.
Also, what is an artificial type? I did not find it in the documentation

Reply:
The method is used during type inference. I renamed the method from `CheckEnds` to `DetermineRootLeaf`, which is more direct. The comment above the method describes what "root" and "leaf" mean in this context.

An "artificial type" refers to `ArtificialType`, as declared in `DafnyAst.cs`. I clarified this in the comment for `DetermineRootLeaf`, and I also expanded on the description for the class in `DafnyAst.cs`.